src/Simulation/QCTraceSimulator/SortedQubitPool.cs

```diff
@@ -0,0 +1,329 @@
+    /// Either time interval or point in time on the timeline of the program.
+    /// </summary>
+    internal struct ComplexTime
+    {
+        /// <summary>
+        /// Tracer configuration interface uses type "double" to provide duration of gates (aka depth).
+        /// When a gate is performed, time advances by the depth of the gate. DepthTime is used to count
+        /// time cosistent with the depth configuration settings. As a result, if a gate time is set to zero,
+        /// DepthTime won't increase after such gate is performed.
+        /// </summary>
+        internal double DepthTime { get; private set; }
+
+        /// <summary>
+        /// TrailingZeroDepthGateCount counts number of zero-depth gates after last non-zero depth gate is performed.
+        /// Generally, gates that happen later on a qubit cannot be scheduled before gates on the same qubit.
+        /// This is tracked transitively via qubit availability time, but zero-depth gates do not allow such calculation.
+        /// The situation is resolved by having this time component - TrailingZeroDepthGateCount.
+        /// </summary>
+        private long TrailingZeroDepthGateCount;
+
+        internal ComplexTime(double time) {
+            DepthTime = time;
+            TrailingZeroDepthGateCount = 0;
+        }
+
+        private ComplexTime(double depthTime, long trailingZeroDepthGateCount) {
+            DepthTime = depthTime;
+            TrailingZeroDepthGateCount = trailingZeroDepthGateCount;
+        }
+
+        public override string ToString() {
+            return $"{DepthTime}d+{TrailingZeroDepthGateCount}";
+        }
+        /// <summary>
+        /// Add gate time to availability time. If gate time is zero, advance dependency time.
+        /// </summary>
+        /// <param name="gateTime">Gate time to advance by. Assumed to be precise for comparison with 0.</param>
+        /// <returns>ComplexTime advanced by provided gate time</returns>
+        internal ComplexTime AdvanceBy(double gateTime) {
+            if (gateTime == 0.0) {
+                return new ComplexTime(DepthTime, TrailingZeroDepthGateCount + 1);
+            }
+            return new ComplexTime(DepthTime + gateTime);
+        }
+
+        /// <summary>
+        /// Subtracts argument ComplexTime from "this" ComplexTime. Returns the result.
+        /// Depth times are assumed to be precise for comparison.
+        /// </summary>
+        /// <param name="time">Time to subtract from this object.</param>
+        /// <returns>Result of subtraction.</returns>
+        internal ComplexTime Subtract(ComplexTime time) {
+            if (DepthTime == time.DepthTime) {
+                return new ComplexTime(0, TrailingZeroDepthGateCount - time.TrailingZeroDepthGateCount);
+            }
+            double result = DepthTime - time.DepthTime;
+            if (result <= 0) {
+                // This could only happen due to insufficient floating point calculation precision.
```

Comment:
Or it can be a logical bug in the tracer, when it tries to subtract a later time from an earlier time...

Reply:
The tracer is bug-free! :) Ha-ha, of course it could happen due to bugs in the tracer but we don't account for this in comments. I'll remove the word 'only'.